### ISSUE TITLE
Polymorphic functions subvert the CEK machine

### ISSUE THREAD

**Issue by effectfully** [input-output-hk/plutus#1882]
With `Bool` in the universe and `ifThenElse` defined like that:

```haskell
typedIfThenElse
    :: (GShow uni, GEq uni, uni `IncludesAll` '[BSL.ByteString, Bool])
    => TypedBuiltinName uni '[Bool, OpaqueTerm uni "a" 0, OpaqueTerm uni "a" 0] (OpaqueTerm uni "a" 0)
typedIfThenElse =
    TypedBuiltinName IfThenElse $
        TypeSchemeAllType Proxy $ \a ->
            Proxy `TypeSchemeArrow` a `TypeSchemeArrow` a `TypeSchemeArrow` TypeSchemeResult a

applyBuiltinName IfThenElse  =
    applyTypedBuiltinName typedIfThenElse (\b x y -> if b then x else y) (\_ _ _ -> ExBudget 1 1)

-- > /\(A :: *) -> \(b : Bool) (x y : () -> A) -> ifThenElse {() -> A} b x y ()
ifThenElse :: (TermLike term TyName Name uni, uni `IncludesAll` '[Bool, ()], Rename (term ())) => term ()
ifThenElse = runQuote $ rename =<< do
    a <- freshTyName () "a"
    b <- freshName () "b"
    x <- freshName () "x"
    y <- freshName () "y"
    let unitFunA = TyFun () unit (TyVar () a)
    return
       . tyAbs () a (Type ())
      $ mkIterLamAbs [
          VarDecl () b bool,
          VarDecl () x unitFunA,
          VarDecl () y unitFunA
          ]
      $ mkIterApp ()
          (tyInst () (builtinNameAsTerm IfThenElse) unitFunA)
          [var () b, var () x, var () y, unitval]
```

and this term

```haskell

appliedIfThenElse :: Term TyName Name DefaultUni ()
appliedIfThenElse =
    let int = mkTyBuiltin @Integer ()
        instConst = Apply () $ mkIterInst () Function.const [int, unit]
    in
            mkIterApp ()
                (TyInst () ifThenElse int)
                [mkConstant () True, instConst $ mkConstant @Integer () 0, instConst $ mkConstant @Integer () 1]
```

which in the readable syntax looks like that:

```
(/\(a_4 :: *) -> \(b_5 : bool) -> \(x_6 : unit -> a_4) -> \(y_7 : unit -> a_4) ->
    ifThenElse {unit -> a_4} b_5 x_6 y_7 ())
  {integer}
  True
  ((/\(a_0 :: *) -> /\(b_1 :: *) -> \(x_2 : a_0) -> \(y_3 : b_1) -> x_2) {integer} {unit} 0)
  ((/\(a_0 :: *) -> /\(b_1 :: *) -> \(x_2 : a_0) -> \(y_3 : b_1) -> x_2) {integer} {unit} 1)
```

the CEK machine fails to evaluate the term:

```
*** Exception: An error has occurred:  Cannot evaluate an open term.
Caused by: (x_2)
```

The following term:

```
(/\(a_4 :: *) -> \(b_5 : bool) -> \(x_6 : unit -> a_0) -> \(y_7 : unit -> a_4) ->
    x_6 ())
  {integer}
  True
  ((/\(a_0 :: *) -> /\(b_1 :: *) -> \(x_2 : a_0) -> \(y_3 : b_1) -> x_2) {integer} {unit} 0)
  ((/\(a_0 :: *) -> /\(b_1 :: *) -> \(x_2 : a_0) -> \(y_3 : b_1) -> x_2) {integer} {unit} 1)
```

which is supposed to behave equally to the previous one doesn't cause the CEK machine to crash.

The original term is handled correctly by the CK machine (evaluation results in `con 0`).

It seems, we can't return an open term from a built-in function application even if that open term is literally one of the arguments of the function, which is super counter-intuitive.

Things to do:

1. add a self-contained test
2. figure out why the CEK machine behaves like it behaves
3. hopefully fix the behaviour

**Comment by effectfully** [comment]
## Problem

Ok, so here's what's going on. Firstly, an MWE is

```
id {integer -> integer} ((\(i : integer) (j : integer) -> i) 1) 0
```

where the meaning of `id` is what you'd expect:

```haskell
dynamicIdMeaning :: DynamicBuiltinNameMeaning uni
dynamicIdMeaning = DynamicBuiltinNameMeaning sch Prelude.id (\_ -> ExBudget 1 1) where
    sch =
        TypeSchemeAllType @"a" @0 Proxy $ \a ->
            a `TypeSchemeArrow` TypeSchemeResult a
```

Here are some intermediate evaluation steps (the variable environment is denoted as `{ <var> :-> (<env>, <value>), ... }`, the context is denoted as a Haskell list of `Frame`s):

<details>
  <summary> Steps </summary>
  <p>

```
{}
[ FrameTyInst (integer -> integer)
, FrameApplyArg {} ((\(i : integer) (j : integer) -> i) 1)
, FrameApplyArg {} 0
] |> id

{}
[ FrameApplyArg {} ((\(i : integer) (j : integer) -> i) 1)
, FrameApplyArg {} 0
] <| id {integer -> integer}

{}
[ FrameApplyFun {} (id {integer -> integer})
, FrameApplyArg {} 0
] |> ((\(i : integer) (j : integer) -> i) 1)

{}
[ FrameApplyArg {} 1
, FrameApplyFun {} (id {integer -> integer})
, FrameApplyArg {} 0
] |> \(i : integer) (j : integer) -> i

{}
[ FrameApplyFun {} (\(i : integer) (j : integer) -> i)
, FrameApplyFun {} (id {integer -> integer})
, FrameApplyArg {} 0
] |> 1

{ i :-> ({}, 1) }
[ FrameApplyFun {} (id {integer -> integer})
, FrameApplyArg {} 0
] |> \(j : integer) -> i

{ i :-> ({}, 1) }
[ FrameApplyFun {} (id {integer -> integer})
, FrameApplyArg {} 0
] |> \(j : integer) -> i
```
</p></details>

It's all fine until we reach

```
{ i :-> ({}, 1) }
[ FrameApplyFun {} (id {integer -> integer})
, FrameApplyArg {} 0
] |> \(j : integer) -> i
```

which transitions into

```
{}
[ FrameApplyArg {} 0
] <| id {integer -> integer} (\(j : integer) -> i)
```

And this is where we have our bug. What happens here is that when we eliminate the function application frame (`FrameApplyFun`) we assume that every argument to a function doesn't have any free variables in it, which is true for all monomorphic functions, because they only receive constants as arguments and constants don't have any variables in them. However a polymorphic function receives an arbitrary term, which may contain free variables.

So right now the machine simply drops the argument's variable environment. This happens [here](https://github.com/input-output-hk/plutus/blob/c7dce437f003e1ef3027762030b6a316bb1cece8/language-plutus-core/src/Language/PlutusCore/Evaluation/Machine/Cek.hs#L268).

## Possible solutions

Can we simply merge the argument's environment into the function's one? Not really, the CEK machine does not rely on the global uniqueness condition and respects the scopes properly. So the scope associated with an argument and the scope associated with a function might have two variables with the same unique that are nonetheless different variables.

Can we somehow restore the argument's environment after evaluation of a built-in function application finishes? Consider

```
... |> choose {integer -> integer -> integer} (\(j : integer) -> i) (\(j : integer) -> k)
```

how do we know whether to restore the environment of the first argument or the second one after `choose` computes? We don't. We'd need to somehow pair every argument with its environment, i.e. either literally pass closures to the constant application machinery (that doesn't know anything about closures, so I don't see how to do that) or use the variable environment in a fancy way.

Here's that fancy way: we take

```
{ i :-> ({}, 1) }
[ FrameApplyFun {} (id {integer -> integer})
, FrameApplyArg {} 0
] |> \(j : integer) -> i
```

and transition it into

```
{ arg :-> (i :-> ({}, 1), \(j : integer) -> i) }
[ FrameApplyArg {} 0
] <| id {integer -> integer} arg
```

i.e. if the argument is not a constant, then we create a new variable, save the old environment in the closure of that variable and apply the function to the variable. If the argument is a constant, then we simply apply the function to it (like before) -- we don't want to introduce any additional overhead in the common case just to handle the less common one.

It might work, but I haven't tried yet. The annoying part is that we somehow need to generate fresh varibles during evaluation.

However we can just discharge the environment into the argument, i.e. instantiate all the free variables. Then we don't need to save the environment. This is what I'm going to implement just to fix the bug and we can think of a better solution later.

**Comment by effectfully** [comment]
> Can we somehow restore the argument's environment after evaluation of a built-in function application finishes? Consider
>
> `... |> choose {integer -> integer -> integer} (\(j : integer) -> i) (\(j : integer) -> k)`
>
> how do we know whether to restore the environment of the first argument or the second one after choose computes?

It's even worse: for a polymorphic built-in `apply` that takes a function and an argument and applies the former to the latter we then somehow have to merge the environments of the arguments. We can't really implement that generally given that the constant application machinery and the machines are completely orthogonal.

And with the

> if the argument is not a constant, then we create a new variable, save the old environment in the closure of that variable and apply the function to the variable.

approach we may pollute the environment with lots of redundant mappings (but it's a general problem with the CEK machine, which we'll likely have to fix at some point as the problem makes our evaluation engine quite more memory-hungry and also less performant than it could be).

But here's a funny thing: if we have saturated builtins then we can always tell inside the CEK machine whether we're evaluating an argument to a lambda or to a builtin. If we additionally insist that all builtins are always eta-expanded, then we can simply not expand variables that map to non-constant terms. I.e. with eta-expanded saturated builtins we don't need to create any new mappings, because they're already there! We just need to stop using them prematurely. The only overhead is that we'll be checking what a variable maps to, but since we'll have to look the variable up in the environment anyway, that's a rather small overhead, and there has to be some overhead anyway. And we preserve the tangentiality of the constant application machines and the machines.

But that will only work reliably if we enforce in the AST the fact that builtins are always applied to variables. I.e. `ApplyBuiltin Builtin [TyName] [Name]` instead of `ApplyBuiltin Builtin [Type] [Term]`. Which prevents us from defining any kind of substitution (because we can't substitute a `Term` for a `Name` in such an AST) and thus breaks the CK machine.

I guess, generating fresh names is just the easiest option.

> effectfully closed this in 3c2654e 11 minutes ago

GitHub trying to be smart.

**Comment by effectfully** [comment]
Now that #1930 is merged, we can probably close that. There might be alternative solutions, but I'm not seeing anything plausible right now and the current one is somewhat satisfying.
